### PR TITLE
Group 3 - Text2SQL Implementation

### PR DESCRIPTION
**Pull request by canonal7**
This pull request introduces several changes, including the addition of SQL table creation and sample data insertion scripts, environment variable definitions, and significant updates to the `RAGHelper.py` file to support database connectivity and text-to-SQL translation. The most important changes are summarized below:

### Environment Variable Definitions:
* Added database connection environment variables (`db_host`, `db_port`, `db_username`, `db_password`, `db_name`) to the `.env.template` file (`server/.env.template`).

### Database Connectivity and Text-to-SQL Translation:
* Introduced PostgreSQL connection and text-to-SQL translation functionalities in `RAGHelper.py`, including methods to connect to the database, retrieve schema, translate natural language queries to SQL, and execute SQL queries (`server/RAGHelper.py`). [[1]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddR30-R33) [[2]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL47-R195) [[3]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL83-R219)

### Code Formatting and Cleanup:
* Reformatted import statements and various methods for better readability and consistency (`server/RAGHelper.py`). [[1]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL5-R18) [[2]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL101-R231) [[3]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL133-R271) [[4]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL174-R309) [[5]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL251-R399) [[6]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL286-R434) [[7]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL302-R457) [[8]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL323-R476) [[9]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL339-R499) [[10]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL359-R512) [[11]](diffhunk://#diff-b34a66842c431a6df3f66798af76838155d3dc7ced67672c18adb84b214f56ddL371-R531)

### CODE REVIEW

**Comment by canonal7** [comment]
### **Report on Pull Request #36: Text-to-SQL Implementation for RAGMeUp Framework**

---

#### **Overview**

The pull request titled **"Group 3 - Text2SQL Implementation"** integrates a **Text-to-SQL (text2sql)** feature into the RAGMeUp framework, enabling users to query structured **PostgreSQL databases** using natural language. This enhancement bridges the gap between intuitive natural language input and SQL-based database querying, expanding the framework's functionality significantly. The pull request includes updates to the backend, user interface, and configuration files to support the new feature.

---

### **Enhancements Introduced**

1. **Text-to-SQL Integration:**
   - Integrated a pre-trained text2sql model that translates natural language queries into SQL queries.
   - Designed the workflow specifically for PostgreSQL databases, leveraging its schema structure and SQL dialect.

2. **Dynamic Schema Retrieval:**
   - Implemented functionality to dynamically fetch database schema details, including table names and their columns, from PostgreSQL.
   - This schema information is formatted and used as context for the text2sql model to ensure accurate SQL generation.

3. **SQL Query Execution:**
   - Added logic to execute SQL queries generated by the text2sql model on a connected PostgreSQL database.
   - Robust error handling ensures meaningful feedback is provided to users for issues like syntax errors or connection failures.

4. **User Interface Enhancements:**
   - Introduced a **toggle button** in the upper-right corner of the UI to allow users to switch between the traditional RAG workflow and the new Text-to-SQL workflow.
   - Enhanced the input handling to process natural language queries for both workflows seamlessly.

5. **Environment Configuration Updates:**
   - Updated the `.env.template` file to include the following PostgreSQL configuration variables:
     - `db_host`: Hostname or IP address of the database server.
     - `db_port`: Port number of the database server.
     - `db_username`: Username for authentication.
     - `db_password`: Password for the database user.
     - `db_name`: Name of the database.
   - Users must populate these variables for the feature to function correctly.

---

### **Specific Code Changes**

1. **Backend Changes:**
   - **`server.py`:**
     - Modified the `/chat` endpoint to accept an `isText2SQL` flag in the request payload, allowing the server to distinguish between workflows.
     - Added logic to call the relevant functions in `RAGHelper` for either RAG or Text-to-SQL workflows.
   - **`RAGHelper.py`:**
     - Added `connect_to_postgres` to establish connections to PostgreSQL databases.
     - Implemented `get_schema` to dynamically retrieve and format schema details.
     - Added `translate_to_sql` for natural language to SQL translation using the text2sql model.
     - Introduced `execute_sql_query` to run generated SQL queries on PostgreSQL and return results.
     - Enhanced `handle_user_interaction` to support both workflows based on the `isText2SQL` flag.
   - **`RAGHelperLocal.py`:**
     - Extended `RAGHelperLocal` to inherit and utilize Text-to-SQL capabilities from `RAGHelper`.

2. **Frontend Changes:**
   - **`index.scala.html`:**
     - Added a toggle button in the UI to switch between workflows, ensuring an intuitive user experience.
     - Adjusted the input field logic to process queries based on the toggle state.
   - **`add.scala.html`:**
     - Updated the layout to align with the new workflow options.
   - **Frontend JavaScript:**
     - Updated event handlers to include the `isText2SQL` flag in requests based on the toggle state.
     - Enhanced query submission logic to process Text-to-SQL queries dynamically.

3. **Environment Configuration:**
   - **`.env.template`:**
     - Added five new variables required for PostgreSQL connectivity:
       - `db_host`, `db_port`, `db_username`, `db_password`, `db_name`.

---

### **Challenges Addressed**

1. **PostgreSQL-Only Focus:**
   - This implementation focuses exclusively on PostgreSQL, leveraging its schema structure and SQL dialect. Adapting the feature for other databases was beyond the current scope.

2. **Dynamic Schema Retrieval:**
   - Fetching and formatting large schemas with complex relationships posed a challenge. The implementation streamlined this by focusing on essential table and column information.

3. **Model Limitations:**
   - The pre-trained text2sql model faced difficulties with ambiguous or incomplete natural language queries, which sometimes led to invalid SQL.

4. **Error Handling:**
   - Comprehensive mechanisms were added to manage common errors during SQL generation and execution, ensuring meaningful feedback for users.

5. **UI Integration:**
   - Adding the toggle feature required careful design to maintain compatibility with existing workflows while providing an intuitive user experience.

---

### **Usage Instructions**

1. **Configure Environment Variables:**
   - Populate the `.env.template` file with the following PostgreSQL details:
     - `db_host`: Hostname or IP (e.g., `localhost`).
     - `db_port`: Port number (default: `5432`).
     - `db_username`: Username for database access.
     - `db_password`: Password for the user.
     - `db_name`: Name of the database.

   Example:
   ```env
   db_host=localhost
   db_port=5432
   db_username=rag_user
   db_password=securepassword
   db_name=rag_db
   ```

2. **Toggle Feature:**
   - Use the toggle button in the upper-right corner of the UI to switch between the RAG and Text-to-SQL workflows.

3. **Start the Application:**
   - Ensure the PostgreSQL database is running and accessible using the provided credentials.
   - Launch the RAGMeUp framework and submit natural language queries via the UI.

---

### **Future Improvements**

1. **Support for Additional Databases:**
   - Expand functionality to support relational databases such as MySQL, SQL Server, and Oracle to enhance versatility.

2. **Connect Databases via UI:**
   - Allow users to input database connection details directly through the UI, reducing reliance on manual environment configuration.

3. **Schema Visualization:**
   - Add a feature to visualize the database schema in the UI, helping users craft precise queries.

4. **Enhanced Error Feedback:**
   - Provide users with detailed feedback, such as the generated SQL query and suggestions for refining inputs when queries fail.

5. **Model Customization:**
   - Fine-tune the text2sql model on domain-specific datasets to improve query accuracy and adaptability.

6. **Caching for Performance:**
   - Implement schema and query result caching to reduce response times for frequently accessed data.

---

### **Conclusion**

This pull request introduces a PostgreSQL-specific Text-to-SQL feature to the RAGMeUp framework, significantly enhancing its ability to handle structured data. The addition of the toggle button in the UI ensures seamless workflow transitions, and the updated configuration provides a streamlined setup process. 

Future work should focus on expanding database compatibility, enabling dynamic database connections via the UI, and refining the user experience to further extend the framework’s usability and versatility. This implementation is a major step forward in making RAGMeUp more robust and user-friendly.
